zebra-consensus/src/checkpoint.rs

```diff
@@ -0,0 +1,158 @@
+};
+use tower::{buffer::Buffer, Service};
+
+use zebra_chain::block::{Block, BlockHeaderHash};
+use zebra_chain::types::BlockHeight;
+
+struct CheckpointVerifier<S> {
+    /// The underlying `ZebraState`.
+    state_service: S,
+
+    /// Each checkpoint consists of a coinbase height and block header hash.
+    ///
+    /// Checkpoints should be chosen to avoid forks or chain reorganizations,
+    /// which only happen in the last few hundred blocks in the chain.
+    /// (zcashd allows chain reorganizations up to 99 blocks, and prunes
+    /// orphaned side-chains after 288 blocks.)
+    checkpoint_list: Option<HashMap<BlockHeight, BlockHeaderHash>>,
+}
+
+/// The error type for the CheckpointVerifier Service.
+// TODO(jlusby): Error = Report ?
+type Error = Box<dyn error::Error + Send + Sync + 'static>;
+
+/// The CheckpointVerifier service implementation.
+///
+/// After verification, blocks are added to the underlying state service.
+impl<S> Service<Arc<Block>> for CheckpointVerifier<S>
+where
+    S: Service<zebra_state::Request, Response = zebra_state::Response, Error = Error>,
+    S::Future: Send + 'static,
+{
+    type Response = BlockHeaderHash;
+    type Error = Error;
+    type Future =
+        Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send + 'static>>;
+
+    fn poll_ready(&mut self, context: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.state_service.poll_ready(context)
+    }
+
+    fn call(&mut self, block: Arc<Block>) -> Self::Future {
+        // TODO(jlusby): Error = Report, handle errors from state_service.
+
+        // These checks are cheap, so we can do them in the call()
+
+        let checkpoints = match &self.checkpoint_list {
+            Some(checkpoints) => checkpoints,
+            None => return async { Err("the checkpoint list is empty".into()) }.boxed(),
+        };
+
+        let block_height = match block.coinbase_height() {
+            Some(height) => height,
+            None => {
+                return async { Err("the block does not have a coinbase height".into()) }.boxed()
+            }
+        };
+
+        // TODO(teor):
+        //   - implement chaining from checkpoints to their ancestors
+        //   - if chaining is expensive, move this check to the Future
+        //   - should the state contain a mapping from previous_block_hash to block?
+        let checkpoint_hash_ref = match checkpoints.get(&block_height) {
+            Some(hash) => hash,
+            None => {
+                return async { Err("the block's height is not a checkpoint height".into()) }
+                    .boxed()
+            }
+        };
+
+        // Avoid moving a reference into the future.
+        let checkpoint_hash = *checkpoint_hash_ref;
+
+        // `state_service.call` is OK here because we already called
+        // `state_service.poll_ready` in our `poll_ready`.
+        let add_block = self.state_service.call(zebra_state::Request::AddBlock {
```

Comment:
Is it safe to call `AddBlock` here prior to doing validation? The future isn't actually awaited until after validation, but this is relying on `state_service` to not commit any data in the body of `call` rather than in the future, correct?

Reply:
Yeah, you're right. in_memory commits in the call, but on_disk commits in the Future.

There's a similar issue in BlockVerifier:
https://github.com/ZcashFoundation/zebra/pull/492/files#diff-7ea86ca07b5d02f05830fd0d71b593f1R59

My understanding is that we want to:
* unconditionally call AddBlock, to avoid filling the buffer with slots reserved by poll_ready()
* do expensive work in the async block
* call AddBlock after the expensive work is done, to avoid committing state early

What's the best way to resolve this issue?

I had trouble moving the AddBlock call into the async block last time I tried. But I can try again.

Or I could write some tests that make sure cancelled futures aren't actually committed?

Reply:
IIUC the only reason that `CheckpointVerifier::call` has to call `self.state_service.call` unconditionally is that it reserved a slot in `poll_ready()`.  But if it didn't reserve a slot, and just returned `Poll::Ready(Ok())` in `poll_ready`, then it could do
```
let mut state = self.state_service.clone();
async move {
    state.ready_and().await?.call(...)
}
```
with no problems -- the issue is just with reserving a slot and then not using it.

Reply:
But we call poll_ready() so we can propagate backpressure?

I don't think that state will be exerting any backpressure yet, so that's probably ok. If ut turns out we need backpressure, it will probably come from the verifiers themselves. Or from the chain reorganisation service.

Reply:
Fixed in 99af8a8966136d5be32748ab6f49e78fe274502c and 086ea49cfce9f9862639f92466805f839908cdd9.